Vislib Point Series updates

ppisljar opened the pull request:
Multiple changes to the Vislib point series charts. Before each of the chart (bar/area/line) would be an independent unit responsible for drawing all of its parts (axes, titles). This change splits things up so we can have greater control and better code reuse.

- converting to ES6
- moving axes out of each chart
- joining both axes types (x and y) into a common axis type 
- allowing multiple axes
- allowing top/bottom/left/right axis positioning
- introducing new chart type 'point_series' which can combine bar/line/area series
- making each of the series (bar/line/area) direction independent (vertical/horizontal charts)

This change should have no impact on the current way Kibana works/looks. All the visualizations should stay the same, no new options were added and the existing ones should keep working just as they did. But this will allow to implement many functionalities that are being asked for a lot like multiple Y axes, switching between line/area/bar chart easily, multiple chart types on one visualization, horizontal bar charts and heatmaps for example.

Fixes:
vislib performs width/height calculations redundantly #6555 
Histogram bar hover only works once #5930 
[vislib] Point series charts hould have bottom line #4067
"Container is too small" flashes briefly #3359

---

For posterity, this pr was submitted in several smaller prs: #8252, #8253, #8254, #8275, #8329, #8950, #8987

cjcenizal commented on src/ui/public/vislib/lib/layout/layout_types.js:
```diff
@@ -13,10 +13,8 @@ export default function LayoutTypeFactory(Private) {
     pie: Private(VislibLibLayoutTypesPieLayoutProvider),
-    tile_map: Private(VislibLibLayoutTypesMapLayoutProvider)
+    tile_map: Private(VislibLibLayoutTypesMapLayoutProvider),
+    point_series: Private(VislibLibLayoutTypesColumnLayoutProvider)
```
Is it safe to assume that point_series is now a base layout type for histogram, line, and area layouts?

ppisljar replied:
yes, even more maybe. point series is now a base chart like pie. area, line and bars are just its series implementations.

cjcenizal commented on src/ui/public/vislib/lib/layout/types/column_layout.js:
```diff
@@ -74,6 +78,27 @@ export default function ColumnLayoutFactory(Private) {
+                    type: 'div',
+                    class: 'x-axis-chart-title',
+                    splits: chartTitleSplit
+                  },*/
```
Can we delete this commented-out code?

cjcenizal commented on src/ui/public/vislib/lib/layout/types/column_layout.js:
```diff
@@ -97,9 +122,43 @@ export default function ColumnLayoutFactory(Private) {
+                    type: 'div',
+                    class: 'y-axis-chart-title',
+                    splits: chartTitleSplit
+                  },*/
```
Can we delete this commented-out code?

cjcenizal commented on src/ui/public/vislib/lib/types/tile_map.js:
```diff
@@ -0,0 +1,19 @@
+      this.charts.forEach(function (chart) {
+        chart.resizeArea();
+      });
+    };*/
```
Can we delete this commented-out code?

cjcenizal commented on src/ui/public/vislib/visualizations/point_series.js:
```diff
@@ -0,0 +1,230 @@
+    };
+
+    draw() {
+      // todo: do we need to handle width and height here ?
```
Might want to uppercase `TODO` here for consistency with other TODOs in the code.

cjcenizal commented on src/ui/public/vislib/visualizations/point_series/_point_seri.js:
```diff
@@ -0,0 +1,87 @@
+
+export default function PointSeriProvider(Private) {
+
+  class PointSeri {
```
Is this name a typo? I don't think "seri" is a word I've ever seen before (and I couldn't find any definition for it online), so the name doesn't help me understand the role of this class. Would just using "series" make sense for the role of this class?

This suggestion applies to the `seri_types.js` file too. That should be `series_types` I think.

spalger replied:
Yeah, `series` is a troublesome word. What's the plural of `series`? `serieses`? it's a bit of a shock at first to me as well, but I think we might be better off just moving away from `series` entirely.

cjcenizal commented:
This is amazing. It's such a huge PR, it's hard for me to wrap my head around so my comments are pretty limited. But from looking at the change in folder structure and how the code has been restructured, I can tell it's a little bit easier to dive into and understand (and scale).

I think two things could really help make vislib more accessible for everyone:

1. **Clarify relationships between files/classes.** This can be greatly improved by changing the import statements to be relative instead of absolute. This will help people trace the dependency graph more easily. Doing this will also help file/folder reorganization to reflect these relationships more clearly.

2. **High-level overview on how vislib works.** I'm just thinking of a simple outline or breakdown of concepts and parts involved in a VIS_LIB.md file somewhere. It's really clear that you have a solid understand of how things work, and it would be great to share that knowledge.

Like I said, though... this is stellar. Thanks for doing this man. LGTM.

ppisljar commented:
thanks @cjcenizal for looking into this. I updated based on your comments. the high level overview will need to get extended but it is a starting point i guess. let me know if you have specific questions you would like me to answer in there.

@spalger was reviewing my work along the way, thanks a lot for this again, so i think now it would be very useful to test this with diff datasets, there are probably things we didn't think off and its very important that this change in deed does work exactly as it is before for the end user.

thomasneirynck commented:
this is obviously a little big to give detailed line-by-line feedback, so I'll try and summarize here:

- I found the structure really legible. I think the design is a seriously good improvement. The addition of external design doc was really helpful to get started.
- Apart from the good typing, the code is a lot cleaner. @spalger @ppisljar you guys rock.
- there's a few todo-tags here in there that I would just remove from the code. This would even include the todos from the old vislib that you preserved. If there is not an external ticket tracking this, they are just going to continue to be mystifying.
- I like the care you took on making API explicit and for example removing external access to "private" variables, and keeping the code flat. That really helps navigating through the code. For example, IntelliJ (and I assume many others) is quite good at type inference as long as you keep the OO simple, and the results are very noticeable with your improvements.

I tested this with existing line, bar-charts, and area-chart visualizations I had saved, and they continued to work. Creating, editing and saving these visualizations was not a problem.

As for whatever version this will go in, I think this needs to go through manual testing. I looked at the existing visualize-section of our tests, and it is already pretty extensive. But specifically, I think we should add manual tests for upgrading between Kibana versions of pre-vislib /post-vislib refactor, which is missing now. Right now, we have similar manual tests for 4->5 upgrades, in the context of the Dashboard. It will be beneficial to have a few specific ones for line/column/area-charts as well solely in the context of Visualize. Perhaps @LeeDr can offer some advice here.

tbragin commented:
@ppisljar From a functional review standpoint, since our goal is to ensure no regressions occur as a result of this PR, in no particular order, I'll use this ticket to mention any differences I see between this PR on top of master and 5.0.1. 

I don't know if any of the changes I notice is due to this PR or master, but nevertheless, here they go:
* I noticed that CSV export has a character in the beginning. I didn't see that in 5.0.1.
<img width="521" alt="screen shot 2016-11-19 at 5 12 52 pm" src="https://cloud.githubusercontent.com/assets/933397/20459581/ee5d7a72-ae7b-11e6-897c-42904f044933.png">
* I noticed one partial bucket to pretty much always be there on the right for any date histogram where data stops. That does not seem to be the case for the same data and visualization on 5.0.1.
<img width="1430" alt="screen shot 2016-11-19 at 5 50 32 pm" src="https://cloud.githubusercontent.com/assets/933397/20459723/b27abb0a-ae80-11e6-920a-2686af8f7f3d.png">
* When selecting "Scaling Y-axis to data bounds", I see some lines cut off at the top.
<img width="1425" alt="screen shot 2016-11-19 at 5 55 45 pm" src="https://cloud.githubusercontent.com/assets/933397/20460070/7831cb12-ae8c-11e6-8890-54181759753b.png">

ppisljar commented:
@tbragin thanks a lot for this !

- CSV export issue I can't reproduce. Does it happen always for you ? or is it with some specific case ? could you please provide exact steps ?
- partial bucket issue: *should be fixed*
- scaling Y axis to data bounds: i can also reproduce this with latest master. the thing is that if you have to points which at a high scale end and you smooth the line it might actually go over the data bounds. we should probably handle that in separate PR (as it seems was not introduced by this change)
- pie chart issues: *should be fixed*
- axis legend cuts off: *should be fixed* i increased the truncate length to 100px.

LeeDr commented:
Is the CSV export weird characters issue related to https://github.com/elastic/kibana/issues/8637

tbragin commented:
@ppisljar 
* CSV export. I just exported the CSV and opened it in Excel for Mac 2011. However, I don't see the special character in vi, and it's not a character I can delete. Did we change anything about the way we generate CSV files as part of this PR?
* Scaling -- Roger, and +1 on tracking as a separate issue and fix in separate PR.
* The rest -- I'll try to pull down again and verify later today.

ppisljar commented:
@LeeDr  yeah it could be. We didnt' change anything around CSV in this PR as far as I'm aware

tbragin commented:
@LeeDr @ppisljar Weird part is that export from 5.0.1 looks fine. But if we didn't change anything about CSV export in this PR, let's not worry about it.

ppisljar commented:
i squashed the commit and rebased on master. let me know if anyone objects. (commits get squashed on merge anyway) ... the original history can still be seen [here](https://github.com/ppisljar/kibana/tree/vislib-axis-refactor-backup)

tbragin commented:
@ppisljar Pulled down latest changes, confirm partial bucket issue, pie chart legend, and legend cut off as addressed.

epixa commented:
https://kibana-ci.elastic.co/view/All/job/elastic+kibana+pull-request+multijob-selenium/958/

epixa commented:
jenkins, test this

ppisljar commented:
I compared with current master using elasticon sample dashboards and beats sample dashboards. Everything seems to be good. Also selenium tests are passing (quite a big part of selenium tests are testing visualize).

thomasneirynck commented:
FWIW, the test-upgrade-dashboard (https://github.com/elastic/kibana-qa/blob/master/exports/kibana4.1.9_export.json).)  also gives following errors:

![image](https://cloud.githubusercontent.com/assets/1833023/20728345/4352f684-b64b-11e6-8ab5-10123a59aa12.png)

This occurs on Master as well though.

thomasneirynck commented:
fyi; github seems to jumble the chronology of these review-comments. This is my latest one: https://github.com/elastic/kibana/pull/9044#pullrequestreview-8513443

ppisljar commented:
i would treat the error on that also appears on master as unrelated (can you create an issue for it and we tackle it separately)

ppisljar commented:
the ` Cannot read property 'toggleClass' of undefined` is the error with vis & no spy panel (which is fixed in master. ... i rebased this and can't reproduce anymore)

cjcenizal commented on src/ui/public/vislib/__tests__/visualizations/area_chart.js:
```diff
@@ -20,12 +20,13 @@ const someOtherVariables = {
 
-_.forOwn(someOtherVariables, function (variablesAreCool, imaVariable) {
-  describe('Vislib Area Chart Test Suite for ' + imaVariable + ' Data', function () {
+_.forOwn(dataTypesArray, function (dataType, dataTypeName) {
```
Thank you for fixing these abominable variable names.... 😉

cjcenizal commented on src/ui/public/vislib/__tests__/visualizations/column_chart.js:
```diff
@@ -58,18 +59,22 @@ dataTypesArray.forEach(function (dataType, i) {
+        if (mode === 'grouped') {
+          expect(isStacked).to.be(false);
+        } else {
+          expect(isStacked).to.be(true);
```
What's the association between the description `append a d.y0 key to the data object` and the assertion, "if mode is grouped, should not be stacked"?

ppisljar replied:
ehm, yeah ... doesnt make much sense ... in the beforeEach i check if d.y0 is there (this means stacking function was run) .... but this test actually tests 'should not stack values if grouped mode is set'

cjcenizal commented on src/ui/public/vislib/__tests__/visualizations/column_chart.js:
```diff
@@ -90,7 +95,7 @@ dataTypesArray.forEach(function (dataType, i) {
+        vis.handler.visConfig.set('mode', 'grouped');
         vis.render(vis.data, persistedState);
       });
 
```
This isn't part of this PR but line 103 has a test with no assertions:

```javascript
      it('should returned grouped bars', function () {
        vis.handler.charts.forEach(function (chart) {});
      });
```

cjcenizal commented on src/ui/public/vislib/__tests__/lib/layout/layout.js:
```diff
@@ -52,22 +56,26 @@ dateHistogramArray.forEach(function (data, i) {
-        expect($(vis.el).find('.x-axis-title').length).to.be(1);
+        expect($(vis.el).find('.x-axis-wrapper').length).to.be(2);
+        expect($(vis.el).find('.x-axis-div-wrapper').length).to.be(2);
+        expect($(vis.el).find('.x-axis-title').length).to.be(2);
```
Where are these numbers coming from? Is there any way to make this test more semantic and robust so that we the reasoning in these tests is clearer?

ppisljar replied:
i don't see an easy way to do it ... but i wouldn't worry about layout class that much as i intend to remove it in 5.3 .... currently we use a combination of HTML and SVG to render charts ... i don't like that, i think it should all be a single SVG element. One direct benefit of that is that it will allow us to save all charts as images. Also i think the code will be clearer and easier to read.

cjcenizal commented on src/ui/public/vislib/__tests__/visualizations/tile_maps/map.js:
```diff
@@ -16,7 +16,7 @@ import VislibVisualizationsMapProvider from 'ui/vislib/visualizations/_map';
 // ];
 
-// // TODO: Test the specific behavior of each these
+//
```
Does this comment still apply to the commented-out code? If so I think we should leave it. If not then can we delete the commented-out code?

ppisljar replied:
i added the comment back. there were generally no changes done to tile maps (except minor things to make it work with the rest of refactor). With Thomas we intend to move maps out of vislib in near future and do some refactoring on it as well. At that time we'll also try to add more extensive tests.

cjcenizal commented on src/ui/public/vislib/__tests__/lib/layout/layout.js:
```diff
@@ -12,6 +12,8 @@ import $ from 'jquery';
 import VislibLibLayoutLayoutProvider from 'ui/vislib/lib/layout/layout';
 import FixturesVislibVisFixtureProvider from 'fixtures/vislib/_vis_fixture';
 import PersistedStatePersistedStateProvider from 'ui/persisted_state/persisted_state';
+import VislibVisConfig from 'ui/vislib/lib/vis_config';
```
Do you want to update these kinds of paths to be relative? I see some of the files have relative paths but some are still absolute like this one.

ppisljar replied:
in the tests i kept the paths absolute to prevent imports like '../lib/vis_config' ... what do you think ?

cjcenizal commented on src/ui/public/vislib/components/zero_injection/inject_zeros.js:
```diff
@@ -19,30 +19,12 @@ export default function ZeroInjectionUtilService(Private) {
-    }
-  }
+  return function (obj, data, orderBucketsBySum = false) {
+    const keys = orderXValues(data, orderBucketsBySum); //_(obj).map('values').flatten().map('x').uniq().value();//
```
Remove commented-out code?

cjcenizal commented on src/ui/public/vislib/lib/axis/axis.js:
```diff
@@ -0,0 +1,330 @@
+      })
+      .offset(this.axisConfig.get('scale.offset', 'zero'));
+
+      if (this.axisConfig.get('scale.mode') === 'stacked'/* && visConfigArgs.type === 'histogram'*/) {
```
Remove commented-out code?

cjcenizal commented on src/ui/public/vislib/lib/axis/axis_config.js:
```diff
@@ -0,0 +1,184 @@
+      defaultYExtents: null,
+      min: null,
+      max: null,
+      mode: 'normal' // [percentage, normal, wiggle, silhouette]
```
Instead of comments like these, we should store these acceptable mode values in a constants file, e.g. `scale_mode.js` and then refer to it here:

```javascript
mode: SCALE_MODE.NORMAL
```

This way the developer knows where to look to see what other kinds of values are acceptable.

cjcenizal commented on src/ui/public/vislib/lib/axis/axis_config.js:
```diff
@@ -0,0 +1,184 @@
+      }
+
+      if (this._values.type === 'value') {
+        const isWiggleOrSilluete = chartConfig.get('mode') === 'wiggle' || chartConfig.get('mode') === 'silhouette';
```
This could refer to `SCALE_MODE.SILHOUETTE`

cjcenizal commented on src/ui/public/vislib/lib/axis/axis_config.js:
```diff
@@ -0,0 +1,184 @@
+      if (this._values.type === 'value') {
+        const isWiggleOrSilluete = chartConfig.get('mode') === 'wiggle' || chartConfig.get('mode') === 'silhouette';
+        // if show was not explicitly set and wiggle or silhouette option was checked
+        if (isWiggleOrSilluete) {
```
This variable name has a typo and should be `isWiggleOrSilhouette`.

cjcenizal commented on src/ui/public/vislib/visualizations/pie_chart.js:
```diff
@@ -2,11 +2,19 @@ import d3 from 'd3';
+    color: undefined,
+    fillColor: undefined,
+    xValue: function (d) { return d.x; },
+    yValue: function (d) { return d.y; }
```
No big deal but these could be terser

```javascript
xValue: d => d.x,
yValue: d => d.y,
```

cjcenizal commented on src/ui/public/vislib/visualizations/point_series/column_chart.js:
```diff
@@ -0,0 +1,251 @@
+        if (d.y < 0) {
+          return Math.abs(yScale(d.y0 + d.y) - yScale(d.y0));
+        }
+        // todo:
```
Should we create an issue and refer to the issue URL instead of an inline comment? I saw Tim doing that and thought it seemed like a good way to make sure these little TODOs don't get lost.

cjcenizal commented on src/ui/public/vislib_vis_type/__tests__/_vislib_renderbot.js:
```diff
@@ -84,15 +84,15 @@ describe('renderbot', function exportWrapper() {
   describe('param update', function () {
-    let params = { one: 'fish', two: 'fish' };
+    let $el = $('<div>testing</div>');
+    let params = { el: $el[0], one: 'fish', two: 'fish' };
```
![image](https://cloud.githubusercontent.com/assets/1238659/20774745/db8724e2-b714-11e6-99e8-ec5e4c04ddb6.png)

w33ble replied:
Fairly certain I wrote the original tests. It pleases me so much that someone noticed this reference 🏆 . It's probably not the only place I've used this either...

thomasneirynck commented:
@ppisljar: fyi, https://github.com/elastic/kibana/issues/9293

cjcenizal commented:
I'm seeing a few errors when I run the browser unit tests:

```
Chrome 54.0.2840 (Mac OS X 10.12.1) tag cloud should only send single renderComplete event: 4 FAILED
	Error: Uncaught Error: expected '23px' to equal '29px' (http://localhost:5610/bundles/tests.bundle.js:294446)
Chrome 54.0.2840 (Mac OS X 10.12.1) tag cloud should show correct data when state-updates are interleaved with resize event FAILED
	Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
Chrome 54.0.2840 (Mac OS X 10.12.1) tag cloud should not put elements in view when container is too small FAILED
	Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
Chrome 54.0.2840 (Mac OS X 10.12.1) tag cloud tags should fit after making container bigger FAILED
	Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
Chrome 54.0.2840 (Mac OS X 10.12.1) tag cloud tags should no longer fit after making container smaller FAILED
	Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
```

ppisljar commented:
@cjcenizal i addressed your comments. Also i found one more bug with mixed data (positive/negative values) on bar charts. This is ready for another look.

cjcenizal commented on src/ui/public/vislib/__tests__/lib/vis_config.js:
```diff
@@ -0,0 +1,113 @@
+
+  describe('get Method', function () {
+    it('should be a function', function () {
+      expect(_.isFunction(visConfig.set)).to.be(true);
```
Not a big deal, but I generally get a little alarm bell when I see assertions against booleans, because failing test messages tend to be less helpful, e.g. `expected true but got false`. Usually they can be structured in a way that will give more info, e.g.

```javascript
expect(typeof visConfig.set).to.be.('function');
```

Will give a message along the lines of `expected 'function' but got 'object'`. I'm OK with the way it is now though, I just wanted to throw the thought out there. Your call. :)

cjcenizal commented on src/ui/public/vislib/lib/axis/axis.js:
```diff
@@ -7,6 +7,7 @@ import AxisLabelsProvider from './axis_labels';
 import AxisScaleProvider from './axis_scale';
 import AxisConfigProvider from './axis_config';
 import errors from 'ui/errors';
+import SCALE_MODES from './scale_modes';
```
Is this being used in this file?

cjcenizal commented on src/ui/public/vislib/lib/axis/scale_modes.js:
```diff
@@ -0,0 +1,9 @@
+  PERCENTAGE: 'percentage',
+  WIGGLE: 'wiggle',
+  SILHOUETTE: 'silhouette',
+  ALL: ['normal', 'percentage', 'wiggle', 'silhouette']
```
`ALL` is a nice touch!

cjcenizal commented on src/ui/public/vislib/lib/dispatch.js:
```diff
@@ -40,7 +40,7 @@ export default function DispatchClass(Private, config) {
       const handler = this.handler;
       const color = _.get(handler, 'data.color');
-      const isPercentage = (handler && handler.visConfig.mode === 'percentage');
+      const isPercentage = (handler && handler.visConfig.get('mode') === 'percentage');
```
Is 'percentage' a `SCALE_MODE`?

thomasneirynck commented:
failure is due to Tag Cloud, this will need to be rebased with Master to include the fix.
